# As a first step, we import pandas, the previous csv file and we drop any columns leftover that we dont intend to use

In [10]:
import pandas as pd

In [11]:
df = pd.read_csv("C:/Users/alexa/OneDrive/Documenten/Collecting Data & Tools and Methods/MomaDatasetVersion1.csv")

In [12]:
df = df.drop(columns=["CreditLine", "AccessionNumber", "DateAcquired", "Cataloged", "ObjectID", "URL", "ThumbnailURL"])

In [13]:
df['Date'].sample(20)

27396                1942-1950
22478                1893-1903
93946                     2005
56779                     1975
111009               June 1981
42250                     1940
105640                    1983
118631                    1929
97507                     1948
98852                  c. 1964
59226     1982-83, dated 1983.
125740               1915-1922
37472                  1904-05
43887                     1923
135090                    1990
83721                  1940–41
52183                     1964
114923                    1960
43847                     1939
96020                     1967
Name: Date, dtype: object

In [14]:
df.dtypes

Unnamed: 0         int64
Title             object
Artist            object
ArtistBio         object
Nationality       object
BeginDate         object
EndDate           object
Gender            object
Date              object
Medium            object
Dimensions        object
Classification    object
Department        object
Wordcount         object
dtype: object

In [15]:
df['Date'].unique

<bound method Series.unique of 0            1896
1            1987
2            1903
3            1980
4            1903
           ...   
138146    c. 1890
138147       1939
138148       1939
138149       1939
138150       1939
Name: Date, Length: 138151, dtype: object>

# This tells us that there are about 138150 rows in the 'date' column, stored as a object type. As you can see above, a problem is that the date rows are formatted in different ways. What we want to do is to clean up the date column, so that we only have one single data format. The way we plan to do that is to remove all punctuation and string characters (both uppercase and lowercase), and to select only the first 4 numbers (which correspond to the first year named). In this way, we select for each artwork the year it was created, or the year at which creation began. 

# In order to do this, we convert the date column into a dictionary, and we store the preliminary results of the data cleaning into a seperate dictionary

In [16]:
dictionary2 = {'Date' : []}

dictionary = df['Date'].to_dict()

In [17]:
def remove_punc(text):
    punctuation = '!@#$%^&*()_-+={}[]:;"\'|<>,.?/~`abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVW '
    return ''.join(character for character in text
                   if character not in punctuation)

for key, value in dictionary.items():
    try:
        dictionary[key] = remove_punc(dictionary[key])
    except:
        dictionary[key] = 'Nan'

for key, value in dictionary.items():
    try: 
        dictionary[key] = dictionary2['Date'].append((dictionary[key])[:4])
    except:
        dictionary[key] = 'Nan'

# As a final step, we store the newly created dictionary as a dataframe, and merge it with the preexisting dataframe from the csv file. The merged dataframes we store as a new csv file. The new csv file is similar to the old csv file, but added to it is the 'data_start_year', which we can subsequently use for the data analysis. 

In [18]:
df2 = pd.DataFrame(dictionary2)

In [19]:
join_df = df.join(df2, lsuffix='', rsuffix='_start_year')

In [20]:
join_df

Unnamed: 0                                              Title  \
0                0  Ferdinandsbrücke Project, Vienna, Austria (Ele...   
1                1  City of Music, National Superior Conservatory ...   
2                2  Villa near Vienna Project, Outside Vienna, Aus...   
3                3  The Manhattan Transcripts Project, New York, N...   
4                4  Villa, project, outside Vienna, Austria, Exter...   
...            ...                                                ...   
138146      138146                                           Untitled   
138147      138147  Plate (folio 2 verso) from Muscheln und schirm...   
138148      138148  Plate (folio 6) from Muscheln und schirme (She...   
138149      138149  Plate (folio 12) from Muscheln und schirme (Sh...   
138150      138150  Plate (folio 16) from Muscheln und schirme (Sh...   

                                                   Artist  \
0                                             Otto Wagner   
1                                Christian de Portzamparc   
2                                              Emil Hoppe   
3                                         Bernard Tschumi   
4                                              Emil Hoppe   
...                                                   ...   
138146  Chesnutt Brothers Studio, Andrew Chesnutt, Lew...   
138147                                 Sophie Taeuber-Arp   
138148                                 Sophie Taeuber-Arp   
138149                                 Sophie Taeuber-Arp   
138150                                 Sophie Taeuber-Arp   

                                          ArtistBio               Nationality  \
0                             (Austrian, 1841–1918)                (Austrian)   
1                               (French, born 1944)                  (French)   
2                             (Austrian, 1876–1957)                (Austrian)   
3         (French and Swiss, born Switzerland 1944)                        ()   
4                             (Austrian, 1876–1957)                (Austrian)   
...                                             ...                       ...   
138146  (American, 1861–1934) (American, 1860–1933)  () (American) (American)   
138147                           (Swiss, 1889–1943)                   (Swiss)   
138148                           (Swiss, 1889–1943)                   (Swiss)   
138149                           (Swiss, 1889–1943)                   (Swiss)   
138150                           (Swiss, 1889–1943)                   (Swiss)   

                BeginDate            EndDate            Gender     Date  \
0                  (1841)             (1918)            (Male)     1896   
1                  (1944)                (0)            (Male)     1987   
2                  (1876)             (1957)            (Male)     1903   
3                  (1944)                (0)            (Male)     1980   
4                  (1876)             (1957)            (Male)     1903   
...                   ...                ...               ...      ...   
138146  (0) (1861) (1860)  (0) (1934) (1933)  () (Male) (Male)  c. 1890   
138147             (1889)             (1943)          (Female)     1939   
138148             (1889)             (1943)          (Female)     1939   
138149             (1889)             (1943)          (Female)     1939   
138150             (1889)             (1943)          (Female)     1939   

                                                   Medium  \
0           Ink and cut-and-pasted painted pages on paper   
1                       Paint and colored pencil on print   
2       Graphite, pen, color pencil, ink, and gouache ...   
3       Photographic reproduction with colored synthet...   
4       Graphite, color pencil, ink, and gouache on tr...   
...                                                   ...   
138146                               Gelatin silver print   
138147  One from an illustrated book

In [21]:
join_df.to_csv("C:/Users/alexa/OneDrive/Documenten/Collecting Data & Tools and Methods/MomaDatasetVersion2.csv")